source/Symbol/SwiftASTContext.cpp

```diff
@@ -8032,3 +8032,173 @@ void lldb_private::printASTValidationInfo(
+                                        bool user_imports, Status &error) {
+  Log *log(lldb_private::GetLogIfAnyCategoriesSet(LIBLLDB_LOG_EXPRESSIONS));
+
+  const std::vector<ConstString> *cu_modules = nullptr;
+  CompileUnit *compile_unit = sc.comp_unit;
+  if (compile_unit && compile_unit->GetLanguage() == lldb::eLanguageTypeSwift)
+    cu_modules = &compile_unit->GetImportedModules();
+  llvm::SmallVector<swift::SourceFile::ImportedModuleDesc, 2>
+      additional_imports;
+
+  std::set<ConstString> loaded_modules;
+  auto load_one_module = [&](const ConstString &module_name) {
+    error.Clear();
+    if (loaded_modules.count(module_name))
+      return true;
+    if (log)
+      log->Printf("[PerformAutoImport] Importing module %s",
+                  module_name.AsCString());
+    loaded_modules.insert(module_name);
+
+    swift::ModuleDecl *swift_module = nullptr;
+    lldb::StackFrameSP this_frame_sp(stack_frame_wp.lock());
+
+    if (module_name == ConstString(this->GetClangImporter()
+                                       ->getImportedHeaderModule()
+                                       ->getName()
+                                       .str()))
+      swift_module = this->GetClangImporter()->getImportedHeaderModule();
+    else if (this_frame_sp) {
+      lldb::ProcessSP process_sp(this_frame_sp->CalculateProcess());
+      if (process_sp)
+        swift_module = this->FindAndLoadModule(module_name, *process_sp.get(),
+                                               error);
+    } else
+      swift_module = this->GetModule(module_name, error);
+
+    if (!swift_module || !error.Success() || this->HasFatalErrors()) {
+      if (log)
+        log->Printf("[PerformAutoImport] Couldn't import module %s: %s",
+                    module_name.AsCString(), error.AsCString());
+      return false;
+    }
+
+    if (log) {
+      log->Printf("Importing %s with source files:", module_name.AsCString());
+
+      for (swift::FileUnit *file_unit : swift_module->getFiles()) {
+        StreamString ss;
+        DescribeFileUnit(ss, file_unit);
+        log->Printf("  %s", ss.GetData());
+      }
+    }
+
+    additional_imports.push_back(swift::SourceFile::ImportedModuleDesc(
+        std::make_pair(swift::ModuleDecl::AccessPathTy(), swift_module),
+        swift::SourceFile::ImportOptions()));
+    return true;
+  };
+
+  if (!user_imports) {
+    if (!load_one_module(ConstString("Swift")))
+      return false;
+    if (cu_modules) {
+      for (const ConstString &module_name : *cu_modules) {
+        if (!load_one_module(module_name))
+          return false;
+      }
+    }
+  } else {
+    llvm::SmallVector<swift::ModuleDecl::ImportedModule, 2> parsed_imports;
+    source_file.getImportedModules(parsed_imports,
+                                   swift::ModuleDecl::ImportFilter::All);
+    auto *persistent_expression_state =
+        sc.target_sp->GetSwiftPersistentExpressionState(exe_scope);
```

Comment:
This is the only bit that seems to be specific to the expression evaluator.

Reply:
PerformAutoImport is used in two different ways, depending on the value of user_imports.

When user_imports is false, then we just look in the current ExecutionContext's CU for modules that that CU imports, and hand import them.  That's because sometimes dependent modules get loaded lazily and we need to force them to be present.

When user_imports is true, then we peer into the swift::SourceFile that is the result of expression parsing and find any import statements that might be there, and import the modules references.  That is to support:

(lldb) expr import Foundation

The first use seems general, but the second use seems very expression parser specific.  I wonder if we oughtn't to split this into two functions, one of which does the general auto-importing and the other handles SourceFile import statements?